### PR TITLE
Feature/emp3 android#11 display kml service data

### PR DESCRIPTION
**Pull request by empbarney**

### CODE REVIEW

**Comment by ee11cbb19052e40b07aac0ca060c23ee** [comment]
see #11

File: mapengine/worldwind/apk/src/main/java/mil/emp3/worldwind/layer/KMLServiceLayer.java [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+import mil.emp3.api.interfaces.IKML;
+import mil.emp3.worldwind.MapInstance;
+import mil.emp3.worldwind.feature.FeatureRenderableMapping;
+import mil.emp3.worldwind.feature.KMLFeature;
+
+/**
+ * KMLService creates KMLFeature(s) and they will get added to this layer.
+ */
+public class KMLServiceLayer extends EmpLayer<IKML> {
+    final static private String TAG = KMLLayer.class.getSimpleName();
```
Comment (ee11cbb19052e40b07aac0ca060c23ee) [inline mapengine/worldwind/apk/src/main/java/mil/emp3/worldwind/layer/KMLServiceLayer.java]: > final static private String TAG = KMLLayer.class.getSimpleName();

should be :
> final static private String TAG = KMLServiceLayer.class.getSimpleName();

File: sdk/sdk-api/src/main/java/mil/emp3/api/KML.java [hunk 2]
```diff
@@ -126,6 +126,18 @@ public KML(InputStream inputStream) throws XmlPullParserException, IOException {
      * @throws IOException This exception is raised in the event of an IO error accessing the URL.
      */
     public KML(java.net.URL url) throws XmlPullParserException, IOException {
+        this(url, null);
+    }
+
+    /**
+     * This constructor creates a KML feature from the KML obtained from the URL.
+     * @param url A valid URL to a KML resource.
+     * @param documentBase A valid filesystem path (e.g. where KMZ fro adding of KML Service was exploded)
```
Comment (ee11cbb19052e40b07aac0ca060c23ee) [inline sdk/sdk-api/src/main/java/mil/emp3/api/KML.java]: Typo:
> where KMZ fro adding of KML Service was exploded

File: sdk/sdk-api/src/main/java/mil/emp3/api/abstracts/MapService.java [hunk 3]
```diff
@@ -21,4 +21,13 @@ protected MapService(String serviceURL)
     public java.net.URL getURL() {
         return this.mapServiceURL;
     }
+
+    @Override
+    public String toString() {
+        if(null != mapServiceURL) {
```
Comment (ee11cbb19052e40b07aac0ca060c23ee) [inline sdk/sdk-api/src/main/java/mil/emp3/api/abstracts/MapService.java:27]: Do we want to provide the class name as part of this string?

File: sdk/sdk-api/src/main/java/mil/emp3/api/utils/kml/EmpKMLParser.java [hunk 4]
```diff
@@ -220,7 +235,26 @@ private void createEMPFeature(EmpObjectHierarchyEntry parentEntry,
...
                 if ((null != kmlStyle) && (null != kmlStyle.getIconUrl())) {
-                    newPoint.setIconURI(kmlStyle.getIconUrl());
+                    if(null == documentBase) {
+                        newPoint.setIconURI(kmlStyle.getIconUrl());
+                    } else {
+                        // documentBase is set by the KMLSProvider when a KMZ file is exploded and stored on local file system.
+                        // If the IconUrl is already a well formed URL then don't touch it. Otherwise it is probably a relative
+                        // path from the KMZ file, so create a file URL for the path.
+                        try {
+                            URL url = new URL(kmlStyle.getIconUrl());
```
Comment (ee11cbb19052e40b07aac0ca060c23ee) [inline sdk/sdk-api/src/main/java/mil/emp3/api/utils/kml/EmpKMLParser.java]: ```url``` is not referenced
> URL url = new URL(kmlStyle.getIconUrl());

File: sdk/sdk-core/aar/src/main/java/mil/emp3/core/services/kml/KMLSProvider.java [hunk 5]
```diff
@@ -0,0 +1,318 @@
...
+        processor = new KMLSProcessor();
+        reporter = new KMLSReporter();
+    }
+
+    /**
+     * Queue up the request for KMLProcessor as we don't want to execute it on the UI thread.
+     * @param map
+     * @param mapService
+     * @param cmrd
+     * @return
```
Comment (ee11cbb19052e40b07aac0ca060c23ee) [inline sdk/sdk-core/aar/src/main/java/mil/emp3/core/services/kml/KMLSProvider.java]: What does the return value signify?

File: sdk/sdk-core/aar/src/main/java/mil/emp3/core/services/kml/KMLSProvider.java [hunk 6]
```diff
@@ -0,0 +1,318 @@
...
+            Log.e(TAG, "addMapService ", e);
+        }
+        return false;
+    }
+
+    /**
+     * This is processed in-line like any other removeFeature.
+     * @param map
+     * @param mapService
+     * @return
```
Comment (ee11cbb19052e40b07aac0ca060c23ee) [inline sdk/sdk-core/aar/src/main/java/mil/emp3/core/services/kml/KMLSProvider.java]: What does the return value signify?

File: sdk/sdk-core/aar/src/main/java/mil/emp3/core/services/kml/KMLSProvider.java [hunk 7]
```diff
@@ -0,0 +1,318 @@
...
+                    // a KML file also.
+                    if((null != request.getKmzDirectory()) && (null != request.getKmzFilePath())) {
+                        try {
+                            ((KMLS) request.getService()).setStatus(request.getMap(), KMLSStatusEnum.EXPLODING);
+                            KMZFile kmzFile = new KMZFile();
+                            kmzFile.unzipKMZFile(request);
+                            reporter.generateEvent(request.getMap(), request.getService(), KMLSEventEnum.KML_SERVICE_FILE_EXPLODED);
+                            request.listFiles(request.getKmzDirectory());
+                        } catch (EMP_Exception e) {
+                            Log.i(TAG, "KMLProcessor-run " + e.getMessage(), e);
```
Comment (ee11cbb19052e40b07aac0ca060c23ee) [inline sdk/sdk-core/aar/src/main/java/mil/emp3/core/services/kml/KMLSProvider.java]: Log.e instead here ?
> Log.i(..

File: sdk/sdk-core/aar/src/main/java/mil/emp3/core/services/kml/KMLSRequest.java [hunk 8]
```diff
@@ -0,0 +1,254 @@
...
+
+    public void setFeature(IKML feature) { this.feature = feature; }
+
+    @Override
+    public UUID getId() {
+        if(null != getService()) {
+            return getService().getGeoId();
+        } else {
+            Log.e(TAG, "getId service is null");
+            return null;
```
Comment (ee11cbb19052e40b07aac0ca060c23ee) [inline sdk/sdk-core/aar/src/main/java/mil/emp3/core/services/kml/KMLSRequest.java]: What are the implications of the service having a null geoId?

File: sdk/sdk-core/aar/src/main/java/mil/emp3/core/services/kml/KMLSRequest.java [hunk 9]
```diff
@@ -0,0 +1,254 @@
...
+            // If target file already exists then no need to fetch it. This can happen if client application
+            // does add/remove/add or there are multiple map instances.
+            if(targetFile.exists() && (targetFile.length() > 0) && (targetFile.canRead())) {
+                Log.i(TAG, "file " + targetFilePath + " exists, no need to fetch");
+            } else {
+
+                // Clean directory just in case and then copy the file over.
+                cleanDirectory(directory);
+
+                try (InputStream input = connection.getInputStream();
```
Comment (ee11cbb19052e40b07aac0ca060c23ee) [inline sdk/sdk-core/aar/src/main/java/mil/emp3/core/services/kml/KMLSRequest.java:185]: Might be able to reduce some code with following:
>  java.nio.file.Files.copy(InputStream, Path, CopyOption...)

File: sdk/sdk-core/aar/src/main/java/mil/emp3/core/storage/MapStatus.java [hunk 10]
```diff
@@ -252,6 +258,15 @@ public boolean hasWMS(java.util.UUID id) {
     }
 
     @Override
+    public boolean serviceExists(UUID uuid) {
+        if(this.mapServiceHash.containsKey(uuid)) {
```
Comment (ee11cbb19052e40b07aac0ca060c23ee) [inline sdk/sdk-core/aar/src/main/java/mil/emp3/core/storage/MapStatus.java]: Can simplify to:
> return this.mapServiceHash.containsKey(uuid);

File: sdk/sdk-core/aar/src/main/java/mil/emp3/core/storage/StorageManager.java [hunk 11]
```diff
@@ -1630,6 +1677,8 @@ public void removeMapService(IMap map, IMapService mapService) throws EMP_Except
         } else {
             oList = new ArrayList<>();
         }
+
+        KMLSRequest.listKmlsRoot();
```
Comment (ee11cbb19052e40b07aac0ca060c23ee) [inline sdk/sdk-core/aar/src/main/java/mil/emp3/core/storage/StorageManager.java]: Maybe add a check for the logging level?
```java
if (Log.isLoggable(TAG, Log.VERBOSE)) {
    KMLSRequest.listKmlsRoot();
}
```

File: sdk/sdk-view/build.gradle [hunk 12]
```diff
@@ -21,6 +21,9 @@ dependencies {
         exclude module: 'hamcrest-core'
         exclude module: 'objenesis'
     }
+
+    testCompile "org.robolectric:robolectric:3.3.2"
```
Comment (ee11cbb19052e40b07aac0ca060c23ee) [inline sdk/sdk-view/build.gradle]: > testCompile "org.robolectric:robolectric:3.3.2"

Declaration of dependency versions should be done in ```common.gradle```.

File: sdk/sdk-core/aar/src/main/java/mil/emp3/core/services/kml/KMLSRequest.java [hunk 13]
```diff
@@ -0,0 +1,254 @@
...
+import mil.emp3.api.interfaces.IKMLS;
+import mil.emp3.api.interfaces.IMap;
+import mil.emp3.api.interfaces.core.storage.IKMLSRequest;
+
+/**
+ * Holds the KML Request till it is removed via removeMapService.
+ *    1. Maintains the current state of the request.
+ *    2. Copies the file to application private folder.
+ */
+public class KMLSRequest implements KMZFile.IKMZFileRerquest, IKMLSRequest {
```
Comment (Ish-rivera) [inline sdk/sdk-core/aar/src/main/java/mil/emp3/core/services/kml/KMLSRequest.java]: Typo IKMZFileRerquest?

File: sdk/sdk-core/aar/src/main/java/mil/emp3/core/services/kml/KMZFile.java [hunk 14]
```diff
@@ -0,0 +1,88 @@
...
+
+/**
+ * Unzip a kmzFile. This is purposly not embedded in KMLRequest as in future there will be a requirement to simply unzip a
+ * specified KMZ file.
+ */
+public class KMZFile {
+    private static String TAG = KMZFile.class.getSimpleName();
+    private final int READ_BUFFER_SIZE = 4096;
+
+    public interface IKMZFileRerquest {
```
Comment (Ish-rivera) [inline sdk/sdk-core/aar/src/main/java/mil/emp3/core/services/kml/KMZFile.java]: Typo IKMZFileRerquest

**Review by Ish-rivera — APPROVED** [review]
Fix the typo

File: tests/test-vandv/src/main/java/mil/emp3/test/emp3vv/navItems/kml_service_test/KMLServiceTest.java [hunk 15]
```diff
@@ -0,0 +1,319 @@
...
+
+    private static String TAG = KMLServiceTest.class.getSimpleName();
+
+    private final StyleManager styleManager;
+
+    private IKMLSEventListener[] kmlsEventListener = new IKMLSEventListener[ExecuteTest.MAX_MAPS];
+
+    String[] sampleURLs = {
+            "https://github.com/downloads/brazzy/nikki/example.kmz",
+            "https://developers.google.com/kml/documentation/KML_Samples.kml"
```
Comment (ee11cbb19052e40b07aac0ca060c23ee) [inline tests/test-vandv/src/main/java/mil/emp3/test/emp3vv/navItems/kml_service_test/KMLServiceTest.java:43]:
```
 "https://github.com/downloads/brazzy/nikki/example.kmz",
 "https://developers.google.com/kml/documentation/KML_Samples.kml"
```
I wonder if we should add these as local testing resources.

**Review by ee11cbb19052e40b07aac0ca060c23ee — CHANGES_REQUESTED** [review]

File: sdk/sdk-core/aar/src/main/java/mil/emp3/core/services/kml/KMLSRequest.java (hunk 13 above)
Comment (empbarney) [reply sdk/sdk-core/aar/src/main/java/mil/emp3/core/services/kml/KMLSRequest.java]: Fixed.

File: mapengine/worldwind/apk/src/main/java/mil/emp3/worldwind/layer/KMLServiceLayer.java (hunk 1 above)
Comment (empbarney) [reply mapengine/worldwind/apk/src/main/java/mil/emp3/worldwind/layer/KMLServiceLayer.java]: Done

File: sdk/sdk-api/src/main/java/mil/emp3/api/KML.java (hunk 2 above)
Comment (empbarney) [reply sdk/sdk-api/src/main/java/mil/emp3/api/KML.java]: Done.

File: sdk/sdk-api/src/main/java/mil/emp3/api/abstracts/MapService.java (hunk 3 above)
Comment (empbarney) [reply sdk/sdk-api/src/main/java/mil/emp3/api/abstracts/MapService.java:27]: No, sub class will print it's own name. MapService is abstract. I updated KMLS toString to call suprt.toString().

File: sdk/sdk-api/src/main/java/mil/emp3/api/utils/kml/EmpKMLParser.java (hunk 4 above)
Comment (empbarney) [reply sdk/sdk-api/src/main/java/mil/emp3/api/utils/kml/EmpKMLParser.java]: Fixed.

File: sdk/sdk-core/aar/src/main/java/mil/emp3/core/services/kml/KMLSProvider.java (hunk 5 above)
Comment (empbarney) [reply sdk/sdk-core/aar/src/main/java/mil/emp3/core/services/kml/KMLSProvider.java]: if service was added then returns true else returns false. updated code.

File: sdk/sdk-core/aar/src/main/java/mil/emp3/core/services/kml/KMLSProvider.java (hunk 7 above)
Comment (empbarney) [reply sdk/sdk-core/aar/src/main/java/mil/emp3/core/services/kml/KMLSProvider.java]: Fixed.

File: sdk/sdk-core/aar/src/main/java/mil/emp3/core/services/kml/KMLSRequest.java (hunk 8 above)
Comment (empbarney) [reply sdk/sdk-core/aar/src/main/java/mil/emp3/core/services/kml/KMLSRequest.java]: This not service having null geoId, this is checking if service itself is null. This should never happen. But I don't want to crash the APP.

File: sdk/sdk-core/aar/src/main/java/mil/emp3/core/services/kml/KMLSRequest.java (hunk 9 above)
Comment (empbarney) [reply sdk/sdk-core/aar/src/main/java/mil/emp3/core/services/kml/KMLSRequest.java:185]: Take a look at this comment on line 126, NIO was ending abruptly, so decided to go the long way, it is working reliably.

// Following would be much more efficient but it is truncating the file. If we have to support
        // Large file sizes then we should investigate the code below.

        // URL website = new URL("http://www.website.com/information.asp");
        // ReadableByteChannel rbc = Channels.newChannel(website.openStream());
        // FileOutputStream fos = new FileOutputStream("information.html");
        // fos.getChannel().transferFrom(rbc, 0, Long.MAX_VALUE);

File: sdk/sdk-core/aar/src/main/java/mil/emp3/core/storage/MapStatus.java (hunk 10 above)
Comment (empbarney) [reply sdk/sdk-core/aar/src/main/java/mil/emp3/core/storage/MapStatus.java]: Done.

File: sdk/sdk-core/aar/src/main/java/mil/emp3/core/storage/StorageManager.java (hunk 11 above)
Comment (empbarney) [reply sdk/sdk-core/aar/src/main/java/mil/emp3/core/storage/StorageManager.java]: Done.

File: sdk/sdk-view/build.gradle (hunk 12 above)
Comment (empbarney) [reply sdk/sdk-view/build.gradle]: Done.

File: tests/test-vandv/src/main/java/mil/emp3/test/emp3vv/navItems/kml_service_test/KMLServiceTest.java (hunk 15 above)
Comment (empbarney) [reply tests/test-vandv/src/main/java/mil/emp3/test/emp3vv/navItems/kml_service_test/KMLServiceTest.java:43]: These specifically hard coded for usability. There is support in VandV to build a File URL from files in /sdcard/testFiles. Additionally one can provide a URL by choosing the "CutomURL" option.

File: sdk/sdk-core/aar/src/main/java/mil/emp3/core/services/kml/KMLSProvider.java (hunk 6 above)
Comment (empbarney) [reply sdk/sdk-core/aar/src/main/java/mil/emp3/core/services/kml/KMLSProvider.java]: Updated.

File: sdk/sdk-core/aar/src/main/java/mil/emp3/core/services/kml/KMZFile.java (hunk 14 above)
Comment (empbarney) [reply sdk/sdk-core/aar/src/main/java/mil/emp3/core/services/kml/KMZFile.java]: Fixed.

**Review by ee11cbb19052e40b07aac0ca060c23ee — APPROVED** [review]

**Comment by mubinakhan** [comment]
Reintegrate

**Pull request merged**